### PR TITLE
allow access to the body from an annotation driven filter

### PR DESCRIPTION
**Pull request by graemerocher**
* allows using regular HTTP binding annotations in filters
* allows binding certain body types and fails otherwise

### CODE REVIEW

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java [hunk 1]
```diff
@@ -45,12 +48,14 @@
  * @author Jonas Konrad
  */
 @Internal
-public final class ImmediateByteBody extends ManagedBody<ByteBuf> implements ByteBody {
+public final class ImmediateByteBody extends ManagedBody<ByteBuf> implements ByteBody, ByteBuffer<ByteBuf> {
```
Comment (yawkat) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: just access the ByteBuf value directly instead and wrap it in NettyByteBuffer, no need to delegate this much

Comment (graemerocher) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: then I would need to couple the filter logic to netty which I can't do the way the current code is structured

Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: i also think that conversion to ByteBuffer should fail if the body is already claimed, it should not be an issue in filters and will prevent race conditions

Comment (graemerocher) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: not sure what you mean here, maybe you want to send a PR with these improvements

File: http/src/main/java/io/micronaut/http/filter/BaseFilterProcessor.java [hunk 2]
```diff
@@ -52,10 +67,50 @@
...
+            public <T> Optional<ArgumentBinder<T, HttpRequest<?>>> findArgumentBinder(Argument<T> argument) {
+                Class<? extends Annotation> annotation = argument.getAnnotationMetadata().getAnnotationTypeByStereotype(Bindable.class).orElse(null);
+                if (annotation != null && PERMITTED_BINDING_ANNOTATIONS.contains(annotation.getName())) {
+                    if (annotation == Body.class) {
+                        return Optional.of((context, source) -> {
+                            if (source instanceof FullHttpRequest<?> fullHttpRequest && fullHttpRequest.isFull()) {
+                                ByteBuffer<?> contents = fullHttpRequest.contents();
+                                if (contents != null) {
+                                    Argument<T> t = context.getArgument();
+                                    if (t.isAssignableFrom(ByteBuffer.class)) {
```
Comment (yawkat) [inline http/src/main/java/io/micronaut/http/filter/BaseFilterProcessor.java:95]: could also go through RawBodyReader for this

Comment (yawkat) [reply http/src/main/java/io/micronaut/http/filter/BaseFilterProcessor.java:95]: advantage would be ByteBuf param support, but maybe that's too difficult to be worth it

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java (hunk 1 above)
Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: sure

Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateByteBody.java]: i think i can make it work without FULL_CONTENT too, i will do it after lunch.

**Review by dstepanov — APPROVED** [review]

**Comment by dstepanov** [comment]
Needs some documentation. Maybe we can have a table of all possible things that can be injected.

**Comment by yawkat** [comment]
@graemerocher i didnt fix all test failures btw

**Review by yawkat — APPROVED** [review]

**Pull request merged**